docs/proposals/cascading-deletion.md

```diff
@@ -0,0 +1,235 @@
+
+```
+type ObjectMeta struct {
+	…
+	Finalizers map[string]string
+}
+```
+
+**ObjectMeta.Finalizers**: the map key is an ad hoc, explanatory name(e.g., "orphan"), the value is an identifier for the responsible component that will remove the finalizer from the map. If the ObjectMeta.DeletionTimestamp is non-nil, entry in the map can only be removed.
+
+## New components
+
+* Finalizer:
+  * Like a controller, a finalizer is always running.
+  * A third party can develop their own finalizer. A finalizer doesn't need to be registered with the API server.
+  * Watches for update events that meet two conditions:
+    1. the updated objects has the identifier of the finalizer in ObjectMeta.Finalizers;
+    2. ObjectMeta.DeletionTimestamp is updated from nil to non-nil.
+  * Applies the finalizing logic to the object in the update event.
+  * Removes itself from ObjectMeta.Finalizers after the finalizing logic is completed.
+  * The API server deletes the object after the last finalizer removes itself from the ObjectMeta.Finalizers field.
+  * Because it's possible for the finalizing logic to be applied multiple times, the finalizing logic has to be idempotent.
+  * If a finalizer fails to act in a timely manner, users can manually remove the finalizer from ObjectMeta.Finalizers by a regular kubectl command.
+
+## Changes to existing components
+
+* API server:
+  * Deletion handler:
+    * If the `ObjectMeta.Finalizers` of the object being deleted is non-empty, then updates the DeletionTimestamp, but does not delete the object.
+    * If the `ObjectMeta.Finalizers` is empty and the [options.GracePeriod](../../pkg/registry/generic/etcd/etcd.go#L405) is zero, then deletes the object. If the options.GracePeriod is non-zero, then just updates the DeletionTimestamp.
+  * Update handler:
+    * If the update removes the last finalizer, and the DeletionTimestamp is non-nil, and the DeletionGracePeriodSeconds is zero, then deletes the object from the registry.
+    * If the update removes the last finalizer, and the DeletionTimestamp is non-nil, but the DeletionGracePeriodSeconds is non-zero, then just updates the object.
+
+## 2B. Enforcing orphaning policty with a finalizer
+
+## API changes
+
+```
+type ObjectMeta struct {
+	...
+	ParentReferences []ObjectReference
+}
+```
+
+**ObjectMeta.ParentReferences**: same as in [design #1](#api-changes).
+
+## New components
+
+* Garbage Collector: Almost identical to the [Garbage Collector](#new-components) in design #1. Some notable differences are
+  * The Garbage Processor doesn't need to remove dangling ParentReferences. The "orphan" finalizer takes care of that.
+  * The Garbage Processor sends a deletion request for an item if none of the object listed in the ParentReferences exist.
+  * Adding a fourth component, the"orphan" finalizer:
+    * Watches for update events as described in [2A](#new-components-1).
+    * Removes the object in the event from the `ParentReferences` of its children.
```

Comment:
Let's leave the namespace out of discussion as we are going to use the existing mechanism to handle it for a while. I'll remove the related sentences from the proposal.

But for the record, in the example you gave, when the orphan finalizer removes RS R's reference to the Deployment, it should also check if R's references become empty, and if so, it adds a reference to the namespace. 
The orphan finalizer won't touch Pod A, since RS R is not being deleted.